File: app/routers/dmca.py
"""Public DMCA endpoints (MOD-002).

Allows rights holders to file claims and content creators to file
counter-notices. Also exposes the designated DMCA agent contact info.
"""
from __future__ import annotations

import logging

from fastapi import APIRouter, Depends, HTTPException

from app.auth.deps import AuthenticatedUser, get_authenticated_user
from app.models import (
    DmcaClaimCreateOut,
    DmcaClaimIn,
    DmcaClaimOut,
    DmcaAgentConfigOut,
    DmcaCounterNoticeIn,
    DmcaCounterNoticeOut,
)
from app.services.dmca_claims import (
    file_counter_notice,
    file_dmca_claim,
    get_claim,
    get_dmca_agent_config,
)

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/v1/dmca", tags=["dmca"])


@router.post("/claims", response_model=DmcaClaimCreateOut, status_code=201)
async def submit_dmca_claim(
    body: DmcaClaimIn,
    user: AuthenticatedUser = Depends(get_authenticated_user),
):
    """Submit a DMCA takedown claim (rights holder)."""
    result = file_dmca_claim(body.model_dump(), claimant_user_id=user.sub)
    return result


@router.get("/claims/{claim_id}", response_model=DmcaClaimOut)
async def get_dmca_claim(
    claim_id: str,
    user: AuthenticatedUser = Depends(get_authenticated_user),
):
    """View a DMCA claim. Only the claimant or the target user may view."""
    item = get_claim(claim_id)
    if not item:
        raise HTTPException(404, "Claim not found")

    # Access control: only claimant or target user
    claimant_user_id = item.get("claimant_user_id", "")
    target_user_id = item.get("target_user_id", "")
    if user.sub not in (claimant_user_id, target_user_id):
        raise HTTPException(403, "Not authorized to view this claim")

    return DmcaClaimOut(
        claim_id=item.get("claim_id", ""),
        status=item.get("status", ""),
        claimant_name=item.get("claimant_name", ""),
        claimant_email=item.get("claimant_email", ""),
        content_url=item.get("content_url", ""),
        content_type=item.get("content_type", ""),
        content_id=item.get("content_id", ""),
        target_user_id=item.get("target_user_id", ""),
        original_work_description=item.get("original_work_description", ""),
        created_at=int(item.get("created_at", 0)),
        updated_at=int(item.get("updated_at", 0)),
        content_removed_at=int(item["content_removed_at"]) if item.get("content_removed_at") else None,
        counter_notice_filed_at=int(item["counter_notice_filed_at"]) if item.get("counter_notice_filed_at") else None,
        waiting_period_expires_at=int(item["waiting_period_expires_at"]) if item.get("waiting_period_expires_at") else None,
        resolved_at=int(item["resolved_at"]) if item.get("resolved_at") else None,
        resolution=item.get("resolution") or None,
        strike_number=int(item.get("strike_number", 0)),
    )


@router.post("/claims/{claim_id}/counter-notice", response_model=DmcaCounterNoticeOut)
async def submit_counter_notice(
    claim_id: str,
    body: DmcaCounterNoticeIn,
    user: AuthenticatedUser = Depends(get_authenticated_user),
):
    """File a DMCA counter-notice (content creator only)."""
    try:
        result = file_counter_notice(claim_id, body.model_dump(), user_id=user.sub)
    except ValueError as exc:
        msg = str(exc)
        if "claim_not_found" in msg:
            raise HTTPException(404, "Claim not found")
        if "invalid_status_transition" in msg:
            raise HTTPException(409, msg)
        raise HTTPException(400, msg)
    except PermissionError:
        raise HTTPException(403, "Only the content creator can file a counter-notice")
    return result


@router.get("/agent-info", response_model=DmcaAgentConfigOut)
async def get_agent_info():
    """Public DMCA designated agent contact information (no auth required)."""
    config = get_dmca_agent_config()
    if not config or not config.get("agent_name"):
        raise HTTPException(404, "DMCA agent information not configured")
    return config

File: frontend/src/api/endpoints/dmca.ts
import { api } from "../client";

// Types
export interface DmcaClaimIn {
  claimant_name: string;
  claimant_email: string;
  claimant_address: string;
  claimant_phone?: string;
  content_url: string;
  content_type: "feed_post" | "feed_media" | "message_media" | "video" | "other";
  content_id?: string;
  original_work_description: string;
  sworn_statement: boolean;
  good_faith_belief: boolean;
  signature: string;
}

export interface DmcaClaimOut {
  claim_id: string;
  status: string;
  claimant_name: string;
  claimant_email: string;
  content_url: string;
  content_type: string;
  content_id: string;
  target_user_id: string;
  original_work_description: string;
  created_at: number;
  updated_at: number;
  content_removed_at?: number;
  counter_notice_filed_at?: number;
  waiting_period_expires_at?: number;
  resolved_at?: number;
  resolution?: string;
  strike_number: number;
}

export interface DmcaClaimCreateOut {
  ok: boolean;
  claim_id: string;
  status: string;
  content_removed: boolean;
  strike_number: number;
  created_at: number;
}

export interface DmcaCounterNoticeIn {
  counter_notice_text: string;
  consent_to_jurisdiction: boolean;
  counter_notice_signature: string;
}

export interface DmcaCounterNoticeOut {
  ok: boolean;
  claim_id: string;
  status: string;
  waiting_period_expires_at: number;
  counter_notice_filed_at: number;
}

export interface DmcaClaimListOut {
  items: DmcaClaimOut[];
  next_cursor?: string;
}

export interface DmcaResolveIn {
  resolution: "restored" | "upheld" | "court_order" | "withdrawn";
  resolution_notes?: string;
}

export interface DmcaResolveOut {
  ok: boolean;
  claim_id: string;
  status: string;
  resolution: string;
  resolved_at: number;
}

export interface DmcaAgentConfig {
  agent_name: string;
  agent_email: string;
  agent_address: string;
  agent_phone?: string;
}

export interface RepeatInfringerStatus {
  user_id: string;
  total_claims: number;
  upheld_claims: number;
  strike_count: number;
  threshold: number;
  status: string;
  claim_history: Array<Record<string, unknown>>;
}

// Public endpoints
export const submitDmcaClaim = (data: DmcaClaimIn) =>
  api.post<DmcaClaimCreateOut>("/v1/dmca/claims", data);

export const getDmcaClaim = (claimId: string) =>
  api.get<DmcaClaimOut>(`/v1/dmca/claims/${claimId}`);

export const fileCounterNotice = (claimId: string, data: DmcaCounterNoticeIn) =>
  api.post<DmcaCounterNoticeOut>(`/v1/dmca/claims/${claimId}/counter-notice`, data);

export const getDmcaAgentInfo = () =>
  api.get<DmcaAgentConfig>("/v1/dmca/agent-info");

// Admin endpoints
export const listDmcaClaims = (params?: Record<string, string>) =>
  api.get<DmcaClaimListOut>("/v1/admin/dmca/claims", params);

export const getDmcaClaimDetail = (claimId: string) =>
  api.get<Record<string, unknown>>(`/v1/admin/dmca/claims/${claimId}`);

export const resolveDmcaClaim = (claimId: string, data: DmcaResolveIn) =>
  api.post<DmcaResolveOut>(`/v1/admin/dmca/claims/${claimId}/resolve`, data);

export const getInfringerStatus = (userId: string) =>
  api.get<RepeatInfringerStatus>(`/v1/admin/dmca/users/${userId}/infringer-status`);

export const updateDmcaAgentConfig = (data: DmcaAgentConfig) =>
  api.put<DmcaAgentConfig>("/v1/admin/dmca/agent-config", data);
